feat: add support for objects containing toPromise method

rafaelss95 opened the pull request:
Closes #1982.

rafaelss95 commented:
@limonte it's still under development. I need some guidance in the tests creation. Could you help me there? BTW, this is more or less the correct flux to implement this?

zenflow commented on sweetalert2.d.ts:
```diff
@@ -337,6 +337,10 @@ declare module 'sweetalert2' {
   }
 
+  interface Promisify<T> {
+    readonly toPromise: () => T;
```
We don't require the `toPromise` member to be `readonly` do we?

rafaelss95 replied:
In fact, `readonly` for this case doesn't matter at all.

zenflow commented on sweetalert2.d.ts:
```diff
@@ -387,7 +391,7 @@ declare module 'sweetalert2' {
   type Awaited<T> = T extends Promise<infer U> ? U : T;
 
-  type SyncOrAsync<T> = T | Promise<T>;
+  type SyncOrAsync<T> = T | Promise<T> | Promisify<T>;
```
I think we can inline `Promisify` here, since it's only used once, and probably will only ever be used once. This makes it easier to see what `SyncOrAsync` is. This also saves us needing to name it.

rafaelss95 replied:
Done.

zenflow commented on src/utils/utils.js:
```diff
@@ -80,4 +80,8 @@ export const warnAboutDepreation = (deprecatedParam, useInstead) => {
+
 export const isPromise = (arg) => arg && Promise.resolve(arg) === arg
+
+export const promisify = (arg) => hasToPromiseFn(arg) ? arg.toPromise() : Promise.resolve(arg)
```
The verb "promisify" has been used a lot to describe converting a callback-based API to a Promise-based API. For example, NodeJS's [util.promisify](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original), Bluebird's [Promise.promisify](http://bluebirdjs.com/docs/api/promise.promisify.html), [pify](https://www.npmjs.com/package/pify), and more. I suggest renaming to something else. Ideas: `toPromise` or `asPromise`.

rafaelss95 replied:
I've updated it to `asPromise`.

zenflow commented:
> I need some guidance in the tests creation. Could you help me there?

@rafaelss95 What in particular do you need guidance on? Are you familiar with automated testing in general? Were you able to run the tests locally?

> BTW, this is more or less the correct flux to implement this?

Not totally sure what you mean by correct "flux".. Can you clarify? Or maybe this doesn't concern me.

rafaelss95 commented:
> Maybe you need help choosing which tests to update?

Yep, I wasn't sure which tests should be updated. I've checked the https://github.com/sweetalert2/sweetalert2#contributing and it doesn't contain anything about the tests.

> Not totally sure what you mean by correct "flux".. Can you clarify? Or maybe this doesn't concern me.

As it's my first contribution to this repository, in addition to `types`, I was asking if the implementation had followed the guidelines.

rafaelss95 commented:
@limonte @zenflow I added some tests for this and I guess it's ready to review now.

PS: Note that I also added tests for `inputOptions` using `Promise` that didn't exist before.

limonte commented:
:tada: This PR is included in version 9.15.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/sweetalert2/v/9.15.0)
- [GitHub release](https://github.com/sweetalert2/sweetalert2/releases/tag/v9.15.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
